// @fast-check/ava/lib/ava-fast-check.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.testProp = exports.test = exports.fc = void 0;
const ava_1 = require("ava");
exports.test = ava_1.default;
const fc = require("fast-check");
exports.fc = fc;
function wrapProp(arbitraries, prop, params) {
    return async (t, ..._args) => {
        var _a;
        let failingTry;
        try {
            await fc.assert(fc.asyncProperty(...arbitraries, async (...args) => {
                const tryResult = await t.try((tt) => prop(tt, ...args));
                if (tryResult.passed) {
                    tryResult.commit();
                    return true;
                }
                const encounteredPreconditionFailure = tryResult.errors.some((error) => fc.PreconditionFailure.isFailure(error.savedError) || fc.PreconditionFailure.isFailure(error.cause));
                if (encounteredPreconditionFailure) {
                    tryResult.discard();
                    fc.pre(false);
                }
                failingTry = tryResult;
                return false;
            }), params);
        }
        catch (error) {
            t.log(error.message);
            ((_a = failingTry === null || failingTry === void 0 ? void 0 : failingTry.commit) !== null && _a !== void 0 ? _a : t.fail)();
        }
        t.pass();
    };
}
function internalTestProp(testFn, label, arbitraries, prop, params) {
    const customParams = Object.assign({}, params);
    if (customParams.seed === undefined) {
        const seedFromGlobals = fc.readConfigureGlobal().seed;
        if (seedFromGlobals !== undefined) {
            customParams.seed = seedFromGlobals;
        }
        else {
            customParams.seed = Date.now() ^ (Math.random() * 0x100000000);
        }
    }
    testFn(`${label} (with seed=${customParams.seed})`, wrapProp(arbitraries, prop, customParams));
}
function exposeModifier(modifier) {
    return (label, arbitraries, prop, params) => internalTestProp(ava_1.default[modifier], label, arbitraries, prop, params);
}
exports.testProp = Object.assign(function testProp(label, arbitraries, prop, params) {
    internalTestProp(ava_1.default, label, arbitraries, prop, params);
}, {
    only: exposeModifier('only'),
    failing: exposeModifier('failing'),
    skip: exposeModifier('skip'),
    serial: exposeModifier('serial'),
    before: ava_1.default.before,
    after: ava_1.default.after,
});
